ui/src/components/setlist/SetlistSongTable.tsx

```diff
@@ -29,156 +29,185 @@ import SetlistChangeKeyModal from './SetlistChangeKeyModal';
+const SetlistSongsTable: React.FC<SetlistSongsTableProps> = ({
+  songList,
+  setSongList,
+  readOnly = false,
+}) => {
+  // State management
+  const [sortedSongList, setSortedSongList] = useState<SongSetlistSchema[]>([]);
+  const [menuState, setMenuState] = useState({
+    anchorEl: null as HTMLElement | null,
+    currentSongId: null as string | null,
+  });
 
+  // Modal states with non-null song when open
+  const [modals, setModals] = useState({
+    preview: {
+      open: false,
+      song: null as SongSetlistSchema | null,
+    },
+    changeKey: {
+      open: false,
+      song: null as SongSetlistSchema | null,
+    },
+  });
+
+  // Sort songs whenever the songList changes
   useEffect(() => {
-    setSortedSongList(songList.sort((a, b) => (a.sequence ? a.sequence - b.sequence : 0)));
+    const sorted = [...songList].sort((a, b) => (a.sequence || 0) - (b.sequence || 0));
+    setSortedSongList(sorted);
   }, [songList]);
 
-  const [anchorEl, setAnchorEl] = useState<null | HTMLElement>(null);
-  const [currentSongId, setCurrentSongId] = useState<string | null>(null);
-
+  // Menu handlers
   const handleMenuOpen = (event: React.MouseEvent<HTMLButtonElement>, songId: string) => {
-    setAnchorEl(event.currentTarget);
-    setCurrentSongId(songId);
+    setMenuState({
+      anchorEl: event.currentTarget,
+      currentSongId: songId,
+    });
   };
 
   const handleMenuClose = () => {
-    setAnchorEl(null);
-    setCurrentSongId(null);
+    setMenuState({
+      anchorEl: null,
+      currentSongId: null,
+    });
   };
 
-  // Song Preview
-  const [songToPreview, setSongToPreview] = useState<SongSetlistSchema | null>(null);
-  const [songPreviewModalOpen, setSongPreviewModalOpen] = useState(false);
-
-  const handleSongPreviewModalOpen = (song: SongSetlistSchema) => {
-    setSongToPreview(song);
-    setSongPreviewModalOpen(true);
+  // Modal handlers with type safety
+  const handleModalToggle = (
+    modalType: 'preview' | 'changeKey',
+    open: boolean,
+    song: SongSetlistSchema | null = null
+  ) => {
+    setModals((prev) => ({
+      ...prev,
+      [modalType]: {
+        open,
+        song: open ? song : null,
+      },
+    }));
   };
-  const handleSongPreviewModalClose = () => {
-    setSongToPreview(null);
-    setSongPreviewModalOpen(false);
-  };
-
-  // Change Key
-  const [songToChangeKey, setSongToChangeKey] = useState<SongSetlistSchema | null>(null);
-  const [changeKeyModalOpen, setChangeKeyModalOpen] = useState(false);
 
-  const handleSongChangeKeyModalOpen = (song: SongSetlistSchema) => {
-    setSongToChangeKey(song);
-    setChangeKeyModalOpen(true);
+  // Type-safe modal open handlers
+  const handlePreviewOpen = (song: SongSetlistSchema) => {
+    handleModalToggle('preview', true, song);
   };
 
-  const handleSongChangeKeyModalClose = () => {
-    setSongToChangeKey(null);
-    setChangeKeyModalOpen(false);
+  const handleChangeKeyOpen = (song: SongSetlistSchema) => {
```

Comment:
[minor]: instad of creating all of this functions, u can just combine all of it by adding a new params for the modalType and pass that function with th enew modalType params to the components.